lib/auth/sessions.go

```diff
@@ -102,6 +106,64 @@ func (s *Server) CreateAppSession(ctx context.Context, req services.CreateAppSes
 	return session, nil
 }
 
+// UpsertAppSession saves the provided app session for the specified user.
+func (s *Server) UpsertAppSession(ctx context.Context, session types.WebSession, identity tlsca.Identity, checker services.AccessChecker) error {
+	// Don't let the app session go longer than the identity expiration.
+	ttl := checker.AdjustSessionTTL(identity.Expires.Sub(s.clock.Now()))
+	session.SetExpiryTime(s.clock.Now().Add(ttl))
+	if err := s.Identity.UpsertAppSession(ctx, session); err != nil {
+		return trace.Wrap(err)
+	}
+	return nil
+}
+
+// WaitForAppSession will block until the requested application session shows up in the
+// cache or a timeout occurs.
+func WaitForAppSession(ctx context.Context, sessionID, user string, ap AccessPoint) error {
```

Comment:
@fspmarshall has been putting functions like these into `services`.

Reply:
This function needs an AccessPoint which resides here in auth package so I can't really move it to services without introducing cyclic deps.

Reply:
You can create an interface with just `GetAppSession` and `NewWatcher` in it.

Reply:
Yeah, I tried that and it doesn't work either unfortunately since this method also depends on services/local so it'd be a loop anyway :-/